### PR TITLE
HIVE-21753: Update HiveMetastore authorization to enable use of HiveA…

### PR DESCRIPTION
**Pull request by rameeshm**
…uthorizer implementation

Signed-off-by: rmani <rmani@hortonworks.com>

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/HiveMetaStoreAuthorizer.java [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("==> HiveMetaStoreAuthorizer.buildAuthzContext(): EventType=" + preEventContext.getEventType());
+    }
+
+    HiveMetaStoreAuthorizableEvent authzEvent = null;
+
+    if (preEventContext != null) {
+      this.preEventContext = preEventContext;
+      switch (preEventContext.getEventType()) {
+        case CREATE_DATABASE:
```
Comment (daijyc) [inline ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/HiveMetaStoreAuthorizer.java]: There are missing events need handling, some of those also require invoking preListener in HiveMetaStore.java
1. create catalog, alter catalog, drop catalog*, admin only
2. drop constrain*, add primary key/foreign key/unique key/not null/default/check constraint*, these require alter table privs
3. exchange partition*, require alter partition privs
4. update/delete table column stats*, require alter table privs
5. update/delete partition column stats*, require alter partition privs
6. create/drop/alter function*, require function privs (if there's any)
7. schema registry related event: add/alter/drop schema, add serde, maybe we can skip for now, but need to track for the future
8. all auth related events: PreAuthorizationCallEvent

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/HiveMetaStoreAuthorizer.java [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+import java.util.List;
+
+/**
+ * HiveMetaStoreAuthorizer :  Do authorization checks on MetaStore Events in MetaStorePreEventListener
+ */
+
+public class HiveMetaStoreAuthorizer extends MetaStorePreEventListener {
+  private static final Log    LOG              = LogFactory.getLog(HiveMetaStoreAuthorizer.class);
+  private static final String HIVE_SUPER_USERS = "hive.metastore.authorization.superusers";
```
Comment (daijyc) [inline ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/HiveMetaStoreAuthorizer.java]: There is already HiveConf.USERS_IN_ADMIN_ROLE, can we reuse?

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/events/AddPartitionEvent.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import java.util.List;
+
+/*
+ Authorizable Event for HiveMetaStore operation  AddPartition
+ */
+
+public class AddPartitionEvent extends HiveMetaStoreAuthorizableEvent {
+  private static final Log LOG = LogFactory.getLog(AddPartitionEvent.class);
+
+  private static final String COMMAND_STR = "alter table add partition";
```
Comment (daijyc) [inline ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/events/AddPartitionEvent.java]: What's the COMMAND_STR used for? Do we need to put table/partition name to it?

File: ql/src/test/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/TestHiveMetaStoreAuthorizer.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+/*
+Test HiveMetaStoreAuthorier
+ */
+public class TestHiveMetaStoreAuthorizer {
```
Comment (daijyc) [inline ql/src/test/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/TestHiveMetaStoreAuthorizer.java]: It's better to instantiate HMSHandler to test this feature. You may set a psudo authorizor in config, and try to do some write operation in HMSHandler.

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/HiveMetaStoreAuthorizableEvent.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import org.apache.hadoop.hive.metastore.api.Database;
+import org.apache.hadoop.hive.metastore.api.StorageDescriptor;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.hadoop.hive.metastore.events.PreEventContext;
+import org.apache.hadoop.hive.ql.security.authorization.plugin.HivePrivilegeObject;
+
+import java.util.Collections;
+import java.util.List;
+
+/*/
```
[the remark this answers is not in the document]
Comment (rameeshm) [reply ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/HiveMetaStoreAuthorizableEvent.java]: Fixed the issue

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/HiveMetaStoreAuthorizer.java [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+  private static PreEventContext preEventContext      = null;
+
+  private String[] superUsers = null;
+
+  private static final ThreadLocal<Configuration> tConfig = new ThreadLocal<Configuration>() {
+      @Override
+      protected Configuration initialValue() {
+    return new HiveConf(HiveMetaStoreAuthorizer.class);
```
[the remark this answers is not in the document]
Comment (rameeshm) [reply ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/HiveMetaStoreAuthorizer.java]: Fixed this issue

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/HiveMetaStoreAuthorizer.java [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+          HiveAuthzSessionContext.Builder authzContextBuilder = new HiveAuthzSessionContext.Builder();
+
+          authzContextBuilder.setClientType(HiveAuthzSessionContext.CLIENT_TYPE.HIVECLI);
+          authzContextBuilder.setSessionString("HiveMetaStore");
+
+          HiveAuthzSessionContext authzSessionContext = authzContextBuilder.build();
+
+          HiveAuthorizer hiveAuthorizer = authorizerFactory.createHiveAuthorizer(new HiveMetastoreClientFactoryImpl(), hiveConf, authenticator, authzSessionContext);
+
+          authenticator.getUserName();
```
[the remark this answers is not in the document]
Comment (rameeshm) [reply ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/HiveMetaStoreAuthorizer.java]: Fixed this issue

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/events/AddPartitionEvent.java (hunk 3 above)
Comment (rameeshm) [reply ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/events/AddPartitionEvent.java]: Fixed this issue

File: ql/src/test/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/TestHiveMetaStoreAuthorizer.java (hunk 4 above)
Comment (rameeshm) [reply ql/src/test/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/TestHiveMetaStoreAuthorizer.java]: Fixed this issue

File: ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/HiveMetaStoreAuthorizer.java (hunk 1 above)
Comment (rameeshm) [reply ql/src/java/org/apache/hadoop/hive/ql/security/authorization/plugin/metastore/HiveMetaStoreAuthorizer.java]: Comment 7 -> add/alter/drop schema -> All these commands will not be authorized for the users 

Comment 8 -> PreAuthorizationCallEvent -> All PreAuthorizationCallEvent call will not be authorized for the users.

Comment  2 to 6 -> all these needs changes to HiveMetastore to fire the auth event and followed authorization changes. I shall address this in subsequent JIRA once this patch is completed and validated. This needs a signification change and verification for it.

Comment 1 -> CREATE CATALOG / ALTER CATALOG / DROP CATALOG can be done as admin user which is already in place. Rest all users will not be authorized to do these operations.

CREATE VIEW / ALTER VIEW / DROP VIEW - operations related to views will not authorized for the users. Only admin users are allowed to do.

**Comment by rameeshm** [comment]
New pull request https://github.com/apache/hive/pull/644

**Pull request closed without merging**